### PR TITLE
Modifies the locations loop to run in parallel batches of 100

### PR DESCRIPTION
**Pull request by groksrc**

### CODE REVIEW

File: csv_to_geo_yaml [hunk 1]
```diff
@@ -94,34 +93,44 @@ addresses.each do |address|
...
+          geo_location = geo_coder_result.data['geometry']['location']
+          result[:address] = geo_coder_result.data['formatted_address']
+          result[:location_type] = geo_coder_result.data['geometry']['location_type']
+          result[:latitude] = geo_location['lat']
+          result[:longitude] = geo_location['lng']
+          result[:search_string] = search_string
+
+          location[:results] << result
+        end
+        lookups_done += 1
```
Comment (mcmoyer) [inline csv_to_geo_yaml]: this is probably ripe for a race condition

File: csv_to_geo_yaml [hunk 2]
```diff
@@ -94,34 +93,44 @@ addresses.each do |address|
...
+          result = {}
+
+          geo_location = geo_coder_result.data['geometry']['location']
+          result[:address] = geo_coder_result.data['formatted_address']
+          result[:location_type] = geo_coder_result.data['geometry']['location_type']
+          result[:latitude] = geo_location['lat']
+          result[:longitude] = geo_location['lng']
+          result[:search_string] = search_string
+
+          location[:results] << result
```
Comment (mcmoyer) [inline csv_to_geo_yaml:119]: this may need to be abstracted out to an object that wraps this in a mutex

File: csv_to_geo_yaml [hunk 3]
```diff
@@ -94,34 +93,44 @@ addresses.each do |address|
...
-    result[:location_type] = geo_coder_result.data['geometry']['location_type']
-    result[:latitude] = geo_location['lat']
-    result[:longitude] = geo_location['lng']
-    result[:search_string] = search_string
-
-    location[:results] << result
+start_time = Time.now
+lookups_done = 0
+batches = locations.each_slice(100).to_a
+batches.each do |batch|
```
Comment (Jakanapes) [inline csv_to_geo_yaml]: any reason this is not just `locations.each_slice(100) do |batch|`?

File: csv_to_geo_yaml [hunk 4]
```diff
@@ -94,34 +93,44 @@ addresses.each do |address|
...
+  threads = []
+  batch.each do |location|
+    next if location[:results].any?
```
Comment (Jakanapes) [inline csv_to_geo_yaml:101]: How big is locations?  Would it make more sense to strip out the entries with any results before looping?

Would it make more sense to do a thread for each batch of 100 rather than looping through batches and doing a thread for each location?

ETA:  Maybe I just don't fully understand the goal of timing the lookups, but it seems like a waste to loop through twice.  Why not just generate `location[:result]` in the addresses loop?  It could still be multithreaded.

File: csv_to_geo_yaml (hunk 1 above)
Comment (Jakanapes) [reply csv_to_geo_yaml]: You could do a thread variable for each count and sum them up before the join.

File: csv_to_geo_yaml (hunk 4 above)
Comment (ruffin) [reply csv_to_geo_yaml:101]: Locations is max 5000, usually several hundred to a few thousand.

I disagree about a thread per batch of 100. The IO in Geocoder.search is the slow part, and single-threaded only gets about 3 per second. , so you'll get lots of speedup per thread. I think this approach would work fine with a mutex on the shared resources.

Comment (Jakanapes) [reply csv_to_geo_yaml:101]: Fair enough, that makes sense.  Still hate to see two loops, though.

Comment (groksrc) [reply csv_to_geo_yaml:101]: I'm open to suggestions on how to improve this. I don't want more than 100 threads at a time, that seems to be the right balance on my workstation anyway.

File: csv_to_geo_yaml (hunk 2 above)
Comment (groksrc) [reply csv_to_geo_yaml:119]: @raelik and I did some testing. We don't believe there's a thread safety problem here. Take a look at the go_fasterer branch that has some changes we made to add thread safety. The results between the two are the same. Adding a mutex effectively makes it single threaded again which defeated the purpose.

**Review by groksrc — COMMENTED** [review]
Please keep in mind that we are trying to retire this script entirely and move the logic into the UI. It's not worth a terrible amount of trouble to get the code perfect. As far as we are aware this PR is an improvement over the existing code and be used immediately.

File: csv_to_geo_yaml (hunk 2 above)
Comment (mcmoyer) [reply csv_to_geo_yaml:119]: That’s the beauty of trying to track down a multi-threaded issue…it may not appear in the testing.

Comment (ruffin) [reply csv_to_geo_yaml:119]: go_fasterer looks safe to me

Comment (joshuawscott) [reply csv_to_geo_yaml:119]: If we are using MRI/YARV then the `<<` operation is atomic, so we should be fine, unless the order of the location[:results] array matters.

**Comment by joshuawscott** [comment]
:shipit:

File: csv_to_geo_yaml (hunk 2 above)
Comment (groksrc) [reply csv_to_geo_yaml:119]: I agree with Blane and Josh. Also, I asked this question generically on SO and got the same answer. https://stackoverflow.com/questions/39516124/ruby-threading-issue-or-no-threading-issue

Comment (raelik) [reply csv_to_geo_yaml:119]: At first glance, this looks problematic because of the `<<`, and I helped Drew change this to avoid using that... but after looking at the rest of the code I realized it was totally unnecessary. 

The important thing to note here is that `location` is always a unique single Hash element of the `locations` array, and that only one thread at a time is ever acting on each of those elements.  The fact that `<<` isn't fully thread-safe isn't actually relevant here, because the actual `<<` operations are happening sequentially within a given thread. `location` isn't shared state, and there is only a shared reference to it held in the `locations` array.  As long as another thread isn't going through that `locations` array and performing operations on the hashes within while these other threads are running, nothing untoward will happen.

Comment (joshuawscott) [inline csv_to_geo_yaml]: actually, `<<` *is* atomic, and that makes all the difference. If it wasn't, then multiple threads could try to insert the `result` into the same position in the array.

File: csv_to_geo_yaml [hunk 5]
```diff
@@ -94,34 +93,43 @@ addresses.each do |address|
...
-    result[:longitude] = geo_location['lng']
-    result[:search_string] = search_string
-
-    location[:results] << result
+start_time = Time.now
+lookups_done = 0
+locations.each_slice(100) do |batch|
+  threads = []
+  batch.each do |location|
+    next if location[:results].any?
```
Comment (joshuawscott) [inline csv_to_geo_yaml:101]: Why is this here?

Comment (mcmoyer) [reply csv_to_geo_yaml:101]: sometimes CS provides us with lat/longs in the csv.  If they do, there are already results and we don’t need to look up any others.

**Comment by mcmoyer** [comment]
🐑 it
